### PR TITLE
ChatAgent OutputFixingParser JSON

### PR DESCRIPTION
**Pull request by jacobrosenthal**
Building on https://github.com/hwchase17/langchainjs/pull/355

Lotta people complaining in discord about agents and chat models. Im still not sure ive seen proof they should be worse, I think its more the current chat model agent is not a great prompt, and to my reading of the original paper doesnt really keep any of the ReAct pattern intact
https://arxiv.org/abs/2210.03629

This brings in the output parser with Fixing. And completely reworks the prompting back to thought/action/input

Curious to hear peoples feedback

### CODE REVIEW

**Comment by jacobrosenthal** [comment]
Oh also improved, print the format_instructions as part of scratch pad, so theyre always reminding the agent right at the point of the next response.

**Comment by jacobrosenthal** [comment]
OH and I like he idea of the finishedToolName being an actual tool presented to the llm directly in order to cut down on the multiple options of responses to one

That said, Ive had this working only off and on so Its cleaner and I like it, but prompts are so brittle when I add my private tools it takes nothing to break it again.. would love to hear others feedback on their use cases or superior prompt engineering

**Comment by jacobrosenthal** [comment]
Cleaned up.  I wanted a test but that seemed broken, also we need something more complex than the olivia wilde thing. Id like to see maybe the sql agent stuff running on something like this

I think a prompt wizard could still clean up more, but mostly working for my use case, and better than what is in master. I could see making a new agent to not break back compat of old one, but leave that up to whomever. 

At least for my use cases it needs the changes to the output parser. I think a benefit of this react style is forcing it to keep all its thoughts in the template, and the current one leaks text outside of the damn code block too often.

new
```
langchain-examples:start: Got output {
langchain-examples:start:   "output": "Harry Styles is currently 28 years old, so the square root of his age is approximately 5.29. Therefore, Olivia Wilde's boyfriend (Harry Styles) is approximately 5.29 years old.",
langchain-examples:start:   "intermediateSteps": [
langchain-examples:start:     {
langchain-examples:start:       "action": {
langchain-examples:start:         "tool": "search",
langchain-examples:start:         "toolInput": "Olivia Wilde boyfriend",
langchain-examples:start:         "log": "```json\n{\n\t\"thought\": \"We need to use the search tool to find out who Olivia Wilde's boyfriend is, and then use the calculator tool to find the square root of his age. We can then use the finished tool to provide the final response.\",\n\t\"tool\": \"search\",\n\t\"input\": \"Olivia Wilde boyfriend\"\n}\n```"
langchain-examples:start:       },
langchain-examples:start:       "observation": "Sudeikis and Wilde's relationship ended in November 2020. Wilde was publicly served with court documents regarding child custody while she was presenting Don't Worry Darling at CinemaCon 2022. In January 2021, Wilde began dating singer Harry Styles after meeting during the filming of Don't Worry Darling."
langchain-examples:start:     },
langchain-examples:start:     {
langchain-examples:start:       "action": {
langchain-examples:start:         "tool": "calculator",
langchain-examples:start:         "toolInput": "sqrt(28)",
langchain-examples:start:         "log": "```json\n{\n\t\"thought\": \"We need to use the calculator tool to find the square root of Harry Styles' age.\",\n\t\"tool\": \"calculator\",\n\t\"input\": \"sqrt(28)\"\n}\n```"
langchain-examples:start:       },
langchain-examples:start:       "observation": "5.291502622129181"
langchain-examples:start:     }
langchain-examples:start:   ]
langchain-examples:start: }
```

vs old
```
langchain-examples:start: Got output {
langchain-examples:start:   "output": "Olivia Wilde's boyfriend is Harry Styles and the square root of his current age is approximately 4.69 (rounded to two decimal places).",
langchain-examples:start:   "intermediateSteps": [
langchain-examples:start:     {
langchain-examples:start:       "action": {
langchain-examples:start:         "tool": "search",
langchain-examples:start:         "toolInput": "Olivia Wilde boyfriend square root of age",
langchain-examples:start:         "log": "{\n    \"action\": \"search\",\n    \"action_input\": \"Olivia Wilde boyfriend square root of age\"\n}"
langchain-examples:start:       },
langchain-examples:start:       "observation": "Olivia Wilde and Harry Styles have been romantically linked for over a year, yet only one is bearing the brunt of criticisms over their age ..."
langchain-examples:start:     }
langchain-examples:start:   ]
langchain-examples:start: }
```

**Comment by jacobrosenthal** [comment]
I went with some wording a little closer to the reflection concept
https://arxiv.org/abs/2303.11366

And Its passing all my (private) tests now. I think im pretty happy with the prompts!

File: examples/src/prompts/structured_parser.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export const run = async () => {
   console.log(input);
   /*
   Answer the users question as best as possible.
-  The output should be a markdown code snippet formatted in the following schema:
+  Your ONLY response should be a fenced code block formatted in the following schema:
```
Comment (jacobrosenthal) [inline examples/src/prompts/structured_parser.ts:30]: "markdown code snippet" has 800k google results, "fenced code block" has 12 million

**Review by hwchase17 — COMMENTED** [review]
sorry for delayed review @jacobrosenthal 

i think this makes sense to me - do you have specific examples of where this failed before? would love love to add those as test cases so we can properly evaluate

File: langchain/src/agents/chat_convo/index.ts [hunk 2]
```diff
@@ -109,13 +119,23 @@ export class ChatConversationalAgent extends Agent {
     for (const step of steps) {
       thoughts.push(new AIChatMessage(step.action.log));
       thoughts.push(
-        new HumanChatMessage(
+        new AIChatMessage(
```
Comment (jacobrosenthal) [inline langchain/src/agents/chat_convo/index.ts:122]: ive gone back and forth on this, making it think it wrote the answer to the tool might occasionally help with it thinking human already knows the results of tools "remember i have forgotten all text"

**Comment by jacoblee93** [comment]
#1292 should address the output fixing parser part - feel free to reopen with prompt changes.

**Pull request closed without merging**